base3-encoding/src/lib.rs

```diff
@@ -0,0 +1,401 @@
+//!
+//! # Encoded Format
+//!
+//! The resulting `Vec<u8>` has a simple structure:
+//! 1.  **Length Prefix (2 bytes)**: A `u16` in little-endian format storing the
+//!     original number of bits (i.e., the length of the input vectors).
+//! 2.  **Data Payload (N * 1 byte)**: A sequence of single bytes, where each
+//!     byte is a `u8` value containing 5 encoded ternary digits.
+
+/// An error that can occur during the encoding process.
+#[derive(Debug, PartialEq, Eq)]
+pub enum EncodeError {
+    /// The provided bit-vectors have unmatching lengths.
+    MismatchedLengths,
+    /// The combination `(true, true)` was found, which is not allowed.
+    InvalidBitCombination,
+    /// The length of the input vectors exceeds `u16::MAX` (65,535).
+    LengthExceedsLimit,
+    /// Arithmetic overflow.
+    ArithmeticOverflow,
+}
+
+// Each u8 chunk can hold 5 base-3 symbols (3^5 = 243 <= 255).
+const BASE3_SYMBOL_PER_CHUNK: usize = 5;
+const ENCODED_BYTES_PER_CHUNK: usize = 1; // std::mem::size_of::<u8>()
+
+/// Encodes two boolean vectors, provided as byte slices, into a single `Vec<u8>`.
+///
+/// # Parameters
+/// - `base_bytes`: The byte slice for the base boolean vector.
+/// - `fallback_bytes`: The byte slice for the fallback boolean vector.
+/// - `num_bits`: The exact number of bits from the start of the slices to encode.
+pub fn encode_from_bytes(
+    base_bytes: &[u8],
+    fallback_bytes: &[u8],
+    num_bits: usize,
+) -> Result<Vec<u8>, EncodeError> {
+    if num_bits > u16::MAX as usize {
+        return Err(EncodeError::LengthExceedsLimit);
+    }
+    let required_bytes = num_bits
+        .checked_add(7)
+        .ok_or(EncodeError::ArithmeticOverflow)?
+        .checked_div(8)
+        .ok_or(EncodeError::ArithmeticOverflow)?;
+
+    if base_bytes.len() < required_bytes || fallback_bytes.len() < required_bytes {
+        return Err(EncodeError::MismatchedLengths);
+    }
+
+    let total_u8_length = num_bits
+        .checked_add(BASE3_SYMBOL_PER_CHUNK - 1)
+        .ok_or(EncodeError::ArithmeticOverflow)?
+        .checked_div(BASE3_SYMBOL_PER_CHUNK)
+        .ok_or(EncodeError::ArithmeticOverflow)?;
+
+    let total_byte_length = total_u8_length
+        .checked_mul(ENCODED_BYTES_PER_CHUNK)
+        .ok_or(EncodeError::ArithmeticOverflow)?;
+
+    let capacity = total_byte_length
+        .checked_add(2)
+        .ok_or(EncodeError::ArithmeticOverflow)?;
+    let mut result = Vec::with_capacity(capacity);
+
+    result.extend_from_slice(&(num_bits as u16).to_le_bytes());
+
+    for chunk_index in 0..total_u8_length {
+        let mut block_num: u8 = 0;
+        let start_bit = chunk_index
+            .checked_mul(BASE3_SYMBOL_PER_CHUNK)
+            .ok_or(EncodeError::ArithmeticOverflow)?;
+        let end_bit = start_bit
+            .checked_add(BASE3_SYMBOL_PER_CHUNK)
+            .ok_or(EncodeError::ArithmeticOverflow)?
+            .min(num_bits);
+
+        for i in (start_bit..end_bit).rev() {
```

Comment:
Another dumb question: is it maybe easier to loop on bit in the outer loop and pack until you hit a full chunk then push the full chunk to the output array?

Reply:
Yeah that is possible. I guess the code would just be a little more verbose since it requires work to manage the state for the chunk manually. The iteration through chunks seem more readable and the compiler should optimize the chunk loop pretty well, but I can explore more if the manual implementation is faster.